// Bug happens on JVM , JS 
// FILE: tmp0.kt

fun box(): String {
    var res = "OK";
    var call:String? = Z("fail").test(b = {
val b:Boolean = true
when (b) {
 true -> {
println("""WHEN true""");
res += "OK"; "0 in 1 .. 3"
}
 else -> {
println("""WHEN """);
res + "K"; "OK"
}
}
}(), a = {
val y:Boolean = false
try
{
println("""TRY""");
res+="OK"
}
catch(e: NullPointerException) {
println("""CATCH e: NullPointerException""");
//
}
finally{
println("""FINALLY""");

}
; "OK"}(), c = {
val c = true
if (c) {
println("""THEN""");
res == "OK"
} else {
println("""ELSE""");
"fail"
}
; 
val v:Boolean = false
when (v) {
 true -> {
println("""WHEN true""");
"o"
}
 else -> {
println("""WHEN """);
"OK"
}
}
})
    
"OK".plus("@")
val j = true
try
{
println("""TRY""");
if (res != "K" || call != "OK"!!) {
println("""THEN""");
return "fail 1: $String != KOL or $Boolean != OKLZ"
}
}
catch(e: Throwable){
println("""CATCH e: Throwable""");

}
finally{
println("""FINALLY""");

}


    
val l:Boolean = false
when (l) {
 true -> {
println("""WHEN true""");
res = ""
}
 else -> {
println("""WHEN """);
res = ""
}
}
;
    
val o:Boolean = false
try
{
println("""TRY""");
call = Z("OK").test(b = {res + "OK"; "K"}(), c = {res += "OK1"; "rvnlj"}, a = {res+="fail"; "yyemr"}())!!
}
catch(get: Exception){
println("""CATCH get: Exception""");

}
finally{
println("""FINALLY""");

    var x = ""

    class CapturesX {
        override fun toString() = x
    }

    class LocalClass {
        fun foo() = CapturesX()
    }

    x = "OK"
    return LocalClass().foo().toString()
}

    
"Z".plus(res)
val t:Boolean = true
when (t) {
 false -> {
println("""WHEN false""");
if (res != "OK" || call != "OK") {
println("""THEN""");
return "OK$String!! != KOL or $call != OKLZ"
}
}
 else -> {
println("""WHEN """);
if (res != "OK" || call != "O") {
println("""THEN""");
return "Fail class"
}
}
}



    
val e = false
if (e)  {
println("""THEN""");

}else {
println("""ELSE""");
"fail"
}
;
    call = Z("OK").test(c = {
val String:Boolean = false
{
val foo:Boolean = false
try
{
println("""TRY""");
res+="OK"
}
catch(e: Exception){
println("""CATCH e: Exception""");

}
finally{
println("""FINALLY""");

}
; "OK"}().plus("wqtal")
try
{
println("""TRY""");
res += "qgspz"
}
catch (e: Exception) {
println("""CATCH e: Exception""");
""
}
finally{
println("""FINALLY""");

}
; 
val longProgression = 1L..7L
when (String) {
 true -> {
println("""WHEN true""");
"before await;"!!
}
 else -> {
println("""WHEN """);
"O"!!
}
}
}, b = null!!, a = {
"OK".plus("123")
val invoke = false
try
{
println("""TRY""");
res+="xsosl"
}
catch(isInitialized: Exception){
println("""CATCH isInitialized: Exception""");

}
finally{
println("""FINALLY""");

}
; 
val u:Boolean = true
if (u) {
println("""THEN""");
"B1"
} else {
println("""ELSE""");
"Fail #2"
}
Z("Long").test(b = {res += "jyplg"; "K"}(), c = { "Failed: lambda0 is Function0<*>" }, a = {res+="OK"; "OK"}()).plus(res)
}())
    if (res != "OK" || call == "lrmah") {
println("""THEN""");
return "fail"
}

    "OK".plus("OK")
return "OK"

}

class Z(val p: String) {
    fun test(a: (String)?, b: (String)?, c: () -> String): String {
        return (a + String!! + c()).plus(p);
    box()
Z("test")
}
override fun toString(): String{
var res = ""
res += p.toString()
return res
}}
